docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+
+* estimation of the current cluster state (node info, available volume claims, etc.)
+* execution of one scheduling iteration
+* update of scheduler caches based on scheduled pod
+
+Some consumed resources do not have to be specified in the pod (e.g. volume claims).
+These kind of consumed resources must be simulated accordingaly.
+
+Once the current state is captured, prediction is no longer dependent on the cluster state.
+Thus, the estimation does not have to correspond to the real scheduling since:
+
+* some nodes can be deleted, some nodes can be evacuated, some nodes can be re-labeled/re-annotated
+* some pods can get evicted in the time of prediction
+* remaining capacity of nodes can increase/decrease
+
+### Other aspects to consider
+
+One could take into account shared resources [3] when predicting free capacity (with fragmenation) as well.
+Currently, limit and requests are supported for cpu and memory resource types only.
+In future, the types can be extended to node disk space resources and custom resource types [4].
+
+Another thing to keep in mind is overcommitment with multiple levels of QoS.
+Or consider compressible vs. incompressible resources.
+
+#### Cluster resource subdivision
+
+Allow cluster resources to be subdivided (https://github.com/kubernetes/kubernetes/issues/442).
+
+#### Resource quota
+
+Resource quota [5] can influence number of pods that can be scheduled per namespace [6] as well.
+Though, this limitation is not hardware specific, it can be taken into account in future implementations.
+
+Example: In a cluster with a capacity of 32 GiB RAM, and 16 cores, let team A use 20 Gib and 10 cores, let B use 10GiB and 4 cores, and hold 2GiB and 2 cores in reserve for future allocation.
+
+#### Federated scheduling
+Out of scope of the document.
+Can be implemented as an agreggation of cluster capacities of individual subclusters.
+
+## Implemenation
+
+Goals:
+
+* Build the implementation over the existing code.
+* Run the final implementation as a self-standing application in a pod (under `kube-system` namespace).
+* Provide REST-API to query questions like "approximately how many more pods could i schedule with this shape?"
+* Integrate the cluster capacity into ``kubectl`` (e.g. ``kubectl cluster-capacity``)
+
+### Code analysis
+
+Currently, each iteration of the scheduler consists of the following [steps](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L93):
+
+1. ask for the next pod ``s.config.NextPod``
+1. schedule the pod ``s.config.Algorithm.Schedule``
+1. update metrics ``metrics.SchedulingAlgorithmLatency.Observe``
+1. update scheduler cache ``s.config.SchedulerCache.AssumePod``
+1. pod binding ``s.config.Binder.Bind``
+
+Scheduling itself (``Schedule``) consists of the following [steps](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L79):
```

Comment:
This is analysis of the current scheduler to show the scheduling algorithms can be abstracted and considered as black boxes. Again, I could convert it into an example and mention it corresponds to the upstream generic scheduler.